client.go

```diff
@@ -167,44 +167,69 @@ func (c *Client) PostForm(endpoint string, vals url.Values, files ...File) (*Res
 			c.addAuth(req)
 			return c.client.Do(req)
 		}
-		body := &bytes.Buffer{}
-		writer := multipart.NewWriter(body)
-
-		// write simple fields
-		for name, vals := range vals {
-			for _, val := range vals {
-				if err := writer.WriteField(name, val); err != nil {
-					return nil, err
-				}
-			}
-		}
 
-		// add files
-		for _, f := range files {
-			w, err := writer.CreateFormFile(f.Name, f.Filename)
-			if err != nil {
-				return nil, err
-			}
-			_, err = io.Copy(w, f.Reader)
+		readBody, writeBody := io.Pipe()
+		// readBody will be closed by client.Do
+
+		writer := multipart.NewWriter(writeBody)
+		errCh := make(chan error, 1)
+		go func() {
+			defer writeBody.Close()
+
+			err := writeForm(writer, vals, files...)
 			if err != nil {
-				return nil, err
+				errCh <- err
+				return
 			}
-		}
-		boundary := writer.Boundary()
-		if err := writer.Close(); err != nil {
-			return nil, err
-		}
-		req, err := http.NewRequest("POST", endpoint, body)
+
+			errCh <- writer.Close()
```

Comment:
the significance of this channel is not clear to me - why not use `writer.CloseWithError(err)` instead? The error should pop up on the other side of the wire:

https://golang.org/pkg/io/#PipeWriter.CloseWithError